calico_rkt/calico_rkt.py

```diff
@@ -20,381 +20,474 @@
+        """Removes the given endpoint from the Calico datastore
 
-    # Remove any IP address assignments that this endpoint has
-    for net in endpoint.ipv4_nets | endpoint.ipv6_nets:
-        assert net.size == 1, "Only 1 address allowed per endpoint. Found in network: %s" % net
-        datastore_client.unassign_address(None, net.ip)
+        :param endpoint:
+        :return: None
+        """
+        try:
+            _log.info("Removing endpoint from the Calico datastore.")
+            self._client.remove_workload(hostname=HOSTNAME,
+                                         orchestrator_id=ORCHESTRATOR_ID,
+                                         workload_id=self.container_id)
+        except KeyError:
+            _log.error("Unable to remove workload with ID %s from datastore.",
+                       self.container_id)
+            sys.exit(1)
 
-    # Remove the endpoint
-    netns.remove_veth(endpoint.name)
+    def _provision_veth(self, endpoint):
+        """Provisions veth for given endpoint.
+
+        Uses the netns relative path passed in through CNI_NETNS_ENV and
+        interface passed in through CNI_IFNAME_ENV.
+
+        :param endpoint
+        :return Calico endpoint object
+        """
+        netns_path = os.path.abspath(os.path.join(os.getcwd(), self.netns))
+        endpoint.mac = endpoint.provision_veth(Namespace(netns_path),
+                                               self.interface)
+        self._client.set_endpoint(endpoint)
+        _log.info("Provisioned veth for endpoint using netns path %s on "
+                  "interface %s", netns_path, self.interface)
+        return endpoint
+
+    def _set_profile_on_endpoint(self, endpoint, profile_name):
+        """Sets given profile_name on given endpoint.
+
+        If profile for given profile_name does not exist, create a profile.
+
+        :param endpoint
+        :param profile_name
+        :return: None.
+        """
+        if not self._client.profile_exists(profile_name):
+            _log.info("Creating new profile %s.", profile_name)
+            self._client.create_profile(profile_name)
+
+        _log.info("Setting profile %s on endpoint.", profile_name)
+        self._client.set_profiles_on_endpoint(profile_names=[profile_name],
+                                              endpoint_id=endpoint.endpoint_id)
 
-    # Remove the container from the datastore.
-    datastore_client.remove_workload(hostname=hostname,
-                                     orchestrator_id=orchestrator_id,
-                                     workload_id=container_id)
+    def _remove_profile(self, profile_name):
+        """Remove the profile from datastore if it has no endpoints attached.
+
+        :param profile_name: Name of profile to remove.
+        :return: None.
+        """
+        profile_members = self._client.get_profile_members(profile_name)
+        if self._client.profile_exists(profile_name) and len(profile_members) < 1:
```

Comment:
As discussed, this won't work because of the race condition when creating / deleting a pod in the same profile at the same time.

We need to solve this, but not in this PR - for now, we just won't delete profiles (there should be very few of them).